File: smart_zambia_invoice/smart_invoice/overrides/backend/item.py
import json

import deprecation

import frappe
import frappe.defaults
from frappe.model.document import Document

from .... import __version__
from ...api.zra_api import make_zra_item_registration
from ...utilities import split_user_mail


@deprecation.deprecated(
    deprecated_in="0.6.2",
    removed_in="1.0.0",
    current_version=__version__,
    details="Use the Register Item button in Item record",
)

def before_insert(doc: Document, method: str) -> None:
    """Item doctype before insertion hook"""
    
    # Default values for tax categories
    vatCatCd = None
    iplCatCd = None
    tlCatCd = None
    exciseTxCatCd = None

    # Check tax type and assign the appropriate category
    tax_type = doc.custom_zra_tax_type

    if tax_type in ["A", "B", "C1", "C2", "C3", "D", "E", "RVAT"]:
        vatCatCd = tax_type

    if tax_type in ["IPL1", "IPL2"]:
        iplCatCd = tax_type

    if tax_type == "TL":
        tlCatCd = tax_type

    if tax_type in ["ECM", "EXEEG"]:
        exciseTxCatCd = tax_type

    # Prepare the registration data
    item_registration_data = {
        "name": doc.name,
        "company_name": frappe.defaults.get_user_default("Company"),
        "itemCd": doc.custom_zra_item_code,
        "itemClsCd": doc.custom_zra_item_classification_code,
        "itemTyCd": doc.custom_product_code,
        "itemNm": doc.item_name,
        "temStdNm": None,
        "orgnNatCd": doc.custom_zra_country_origin_code,
        "pkgUnitCd": doc.custom_zra_packaging_unit_code,
        "qtyUnitCd": doc.custom_zra_unit_quantity_code,
        "vatCatCd": vatCatCd,
        "iplCatCd": iplCatCd,
        "tlCatCd": tlCatCd,
        "exciseTxCatCd": exciseTxCatCd,
        "btchNo": None,
        "bcd": None,
        "dftPrc": f"{doc.valuation_rate:.2f}" if doc.valuation_rate else "0.00",
        "addInfo": None,
        "sftyQty": None,
        "manufactuterTpin": None,
        "manufacturerItemCd": None,
        "rrp": doc.standard_rate or 0.00,  # Use field if set, else default
        "svcChargeYn": "Y",
        "rentalYn": "N",
        "useYn": "Y",
        "regrId": split_user_mail(doc.owner),
        "regrNm": doc.owner,
        "modrId": split_user_mail(doc.modified_by),
        "modrNm": doc.modified_by,
    }

    # Make the ZRA item registration request
    make_zra_item_registration(json.dumps(item_registration_data))


def validate(doc: Document, method: str) -> None:
    """
    Generates a unique item code for the CIS compliance.
    Format: {Country of Origin}{Product Type}{Packaging Unit}{Quantity Unit}{Incremented Number}
    Example: ZM2NTBA0000012
    """
    # Step 1: Validate required fields for prefix generation
    required_fields = {
        "Country of Origin Code": doc.custom_zra_country_of_origin,
        "Product Type Code": doc.custom_zra_product_type_code,
        "Packaging Unit Code": doc.custom_zra_packaging_unit,
        "Quantity Unit Code": doc.custom_zras_unit_of_quantity,
    }

    missing_fields = [key for key, value in required_fields.items() if not value]
    if missing_fields:
        frappe.throw(f"Missing required fields for item code generation: {', '.join(missing_fields)}")

    # Step 2: Generate the prefix
    prefix = (
        f"{doc.custom_zra_country_origin_code}"
        f"{doc.custom_product_code}"
        f"{doc.custom_zra_packaging_unit}"
        f"{doc.custom_zras_unit_of_quantity}"
    )

    # Step 3: Check if the item already has a code and extract suffix if present
    if doc.custom_zra_item_code and doc.custom_zra_item_code.startswith(prefix):
        # Extract the numeric suffix
        suffix = doc.custom_zra_item_code[-7:]
    else:
        # Query the last item code with the same prefix
        last_code = frappe.db.sql(
            """
            SELECT custom_zra_item_code
            FROM `tabItem`
            WHERE custom_zra_item_code LIKE %s
            ORDER BY CAST(SUBSTRING(custom_zra_item_code, -7) AS UNSIGNED) DESC
            LIMIT 1
            """,
            (f"{prefix}%",),
            as_dict=True,
        )

        # Increment the suffix or start from '0000001'
        if last_code:
            last_suffix = int(last_code[0]["custom_zra_item_code"][-7:])
            suffix = str(last_suffix + 1).zfill(7)
        else:
            suffix = "0000001"

    # Step 4: Combine the prefix and suffix to form the complete item code
    doc.custom_zra_item_code = f"{prefix}{suffix}"

    # Optional: Log the generated code for audit
    frappe.logger().info(
        f"Generated item code {doc.custom_zra_item_code} for item {doc.name}"
    )


@frappe.whitelist()
def prevent_item_deletion(doc, method):
    if doc.custom_item_registered == 1:
        frappe.throw(_("Cannot delete registered items"))

File: smart_zambia_invoice/smart_invoice/overrides/frontend/item.js
const itemDoctypName = "Item";

frappe.ui.form.on(itemDoctypName, {
  refresh: function (frm) {
    const companyName = frappe.boot.sysdefaults.company;

    // if (frm.doc.custom_zra_item_registered_) {
    //   frm.toggle_enable("custom_zra_item_classification_code", false);
    //   frm.toggle_enable("custom_zra_country_of_origin", false);
    //   frm.toggle_enable("custom_zra_tax_type", false);
    //   frm.toggle_enable("custom_zra_packaging_unit", false);
    //   frm.toggle_enable("custom_zras_unit_of_quantity", false);
    //   frm.toggle_enable("custom_zra_product_type_code", false);
    // }

    // if (frm.doc.custom_zra_imported_item_submitted_) {
    //   frm.toggle_enable("custom_zra_referenced_imported_item", false);
    //   frm.toggle_enable("custom_zra_imported_item_status", false);
    // }

    if (!frm.is_new()) {
      if (
        !frm.doc.custom_zra_item_registered_ &&
        !frm.doc.custom_has_a_recommended_retail_price_rrp_
      ) {
        frm.add_custom_button(
          __("Register/Update Item"),
          function () {
            let taxType = frm.doc.custom_zra_tax_type || null;
            let vatCatCd = null;
            let iplCatCd = null;
            let tlCatCd = null;
            let exciseTxCatCd = null;

            // VAT Category Mapping
            if (
              ["A", "B", "C1", "C2", "C3", "D", "E", "RVAT"].includes(taxType)
            ) {
              vatCatCd = taxType;
            }

            // Insurance Premium Levy Mapping
            if (["IPL1", "IPL2"].includes(taxType)) {
              iplCatCd = taxType;
            }

            if (taxType === "TOT") {
              tlCatCd = taxType;
            }

            // Tourism Levy Mapping
            if (["TL", "F"].includes(taxType)) {
              tlCatCd = taxType;
            }

            // Excise Tax Mapping
            if (["ECM", "EXEEG"].includes(taxType)) {
              exciseTxCatCd = taxType;
            }

            frappe.call({
              method:
                "smart_zambia_invoice.smart_invoice.api.zra_api.make_zra_item_registration",
              args: {
                request_data: {
                  name: frm.doc.name,
                  company_name: companyName,
                  itemCd: frm.doc.custom_zra_item_code,
                  itemClsCd: frm.doc.custom_zra_item_classification_code,
                  itemTyCd: frm.doc.custom_product_code,
                  itemNm: frm.doc.item_name,
                  itemStdNm: frm.doc.item_name,
                  orgnNatCd: frm.doc.custom_zra_country_origin_code,
                  pkgUnitCd: frm.doc.custom_zra_packaging_unit_code,
                  qtyUnitCd: frm.doc.custom_zra_unit_quantity_code,
                  vatCatCd: vatCatCd,
                  iplCatCd: iplCatCd,
                  tlCatCd: tlCatCd,
                  exciseTxCatCd: exciseTxCatCd,
                  btchNo: null,
                  bcd: null,
                  dftPrc: frm.doc.valuation_rate.toFixed(2),
                  addInfo: null,
                  sftyQty: "0",
                  manufactuterTpin: "1000000000",
                  manufacturerItemCd: "ZM2EA1234",
                  rrp: frm.doc.custom_recommended_retail_price,
                  svcChargeYn: "Y",
                  rentalYn: "N",
                  useYn: "Y",
                  regrId: frm.doc.owner,
                  regrNm: frm.doc.owner,
                  modrId: frm.doc.modified_by,
                  modrNm: frm.doc.modified_by,
                },
              },
              callback: (response) => {
                frappe.msgprint(
                  "Item Registration Queued. Please refresh the browser tab."
                );
              },
              error: (error) => {
                // Error Handling is Deferred to the Server
              },
            });
          },
          __("ZRA Actions")
        );
      }
      if (frm.doc.is_stock_item) {
        frm.add_custom_button(
          __("ZRA Item Inventory Submission"),
          function () {
            frappe.call({
              method:
                "smart_zambia_invoice.smart_invoice.api.zra_api.save_stock_inventory",
              args: {
                request_data: {
                  company_name: companyName,
                  name: frm.doc.name,
                  itemName: frm.doc.item_code,
                  remainq: frm.doc.stock_levels,
                  itemCd: frm.doc.custom_zra_item_classification_code,
                  registered_by: frm.doc.owner,
                  modified_by: frm.doc.modified_by,
                },
              },
              callback: (response) => {},
              error: (error) => {
                // Error Handling is Defered to the Server
              },
            });
          },
          __("ZRA Actions")
        );
      }
      if (
        frm.doc.custom_has_a_recommended_retail_price_rrp_ &&
        !frm.doc.custom_zra_item_registered_
      ) {
        frm.add_custom_button(
          __("Submit RRP Item"),
          function () {
            let originalItemName = frm.doc.item_name || "";

            // Ensure item name is a valid string and trim if necessary
            let shortenedItemName = originalItemName.substring(0, 200);

            frappe.call({
              method:
                "smart_zambia_invoice.smart_invoice.api.zra_api.make_rrp_item_registration",
              args: {
                request_data: {
                  name: frm.doc.name,
                  company_name: companyName,
                  itemCd: frm.doc.custom_zra_item_code,
                  itemClsCd: frm.doc.custom_zra_item_classification_code,
                  itemTyCd: frm.doc.custom_product_code,
                  itemNm: shortenedItemName, 
                  taxTyCd: frm.doc.custom_zra_tax_type,
                  orgnNatCd: frm.doc.custom_zra_country_origin_code,
                  pkgUnitCd: frm.doc.custom_zra_packaging_unit_code,
                  qtyUnitCd: frm.doc.custom_zra_unit_quantity_code,
                  rrp: frm.doc.standard_rate,
                  useYn: "Y",
                  regrId: frm.doc.owner,
                  regrNm: frm.doc.owner,
                  modrId: frm.doc.modified_by,
                  modrNm: frm.doc.modified_by,
                },
              },
              callback: (response) => {
                frappe.msgprint("Request queued. Please Refresh your tab");
              },
              error: (error) => {
                console.error("Error:", error);
              },
            });
          },
          __("ZRA Actions")
        );
      }

      // if (!frm.doc.custom_zra_item_registered_) {
      //   frm.add_custom_button(
      //     __("Submit Imported Item"),
      //     function () {
      //       frappe.call({
      //         method:
      //           "smart_zambia_invoice.smart_invoice.api.zra_api.make_zra_item_registration",
      //         args: {
      //           request_data: {
      //             company_name: companyName,
      //             name: frm.doc.name,
      //             item_sequence: frm.doc.idx,
      //             item_code: frm.doc.custom_zra_item_code,
      //             task_code: frm.doc.custom_zra_imported_item_task_code,
      //             item_classification_code:frm.doc.custom_zra_item_classification_code,
      //             import_item_status:frm.doc.custom_zra_imported_item_status_code,
      //             hs_code: frm.doc.custom_zra_hs_code,
      //             modified_by: frm.doc.modified_by,
      //             declaration_date: frm.doc.creation,
      //           },
      //         },
      //         callback: (response) => {
      //           frappe.msgprint("Request queued. Check later.");
      //         },
      //         error: (error) => {
      //           // Error Handling is Defered to the Server
      //         },
      //       });
      //     },
      //     __("ZRA Actions")
      //   );
      // }
    }
  },
  custom_product_type_name: function (frm) {
    if (frm.doc.custom_product_type_name === "Service") {
      frm.set_value("is_stock_item", 0);
    } else {
      frm.set_value("is_stock_item", 1);
    }
  },
});
